File: lib/storage.ts
// lib/storage.ts
//
// Safe localStorage wrapper. Guards against SSR (no window) and
// sandboxed/blocked storage (access throws). Falls back to an
// in-memory map so the app never crashes. On a normal browser
// session everything persists across reloads and tab closes.

import type {
  Bottle,
  Transaction,
  Expense,
  BusinessSettings,
  DEFAULT_BUSINESS_SETTINGS,
  BackupData,
} from "./types";

const mem: Record<string, string> = {};

function canUseLS(): boolean {
  try {
    return typeof window !== "undefined" && !!window.localStorage;
  } catch {
    return false;
  }
}

export function getItem(key: string): string | null {
  try {
    if (canUseLS()) return window.localStorage.getItem(key);
  } catch {
    /* fall through */
  }
  return key in mem ? mem[key] : null;
}

export function setItem(key: string, value: string): void {
  try {
    if (canUseLS()) {
      window.localStorage.setItem(key, value);
      return;
    }
  } catch {
    /* fall through */
  }
  mem[key] = value;
}

export function removeItem(key: string): void {
  try {
    if (canUseLS()) {
      window.localStorage.removeItem(key);
      return;
    }
  } catch {
    /* fall through */
  }
  delete mem[key];
}

/* ── GSP ── */
export const LS_GSP = "gasrefill_gsp_v1";

export function loadGsp(fallback: string): string {
  const v = getItem(LS_GSP);
  return v !== null && v !== "" ? v : fallback;
}

export function saveGsp(v: string): void {
  setItem(LS_GSP, v);
}

/* ── History ── */
export const LS_HIST = "gasrefill_history_v1";

export function loadHistory(): Transaction[] {
  try {
    const raw = getItem(LS_HIST);
    const parsed = raw ? (JSON.parse(raw) as Transaction[]) : [];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

export function saveHistory(h: Transaction[]): void {
  setItem(LS_HIST, JSON.stringify(h));
}

/* ── Bottles ── */
export const LS_BOTTLES = "gasrefill_bottles_v1";

export function loadBottles(): Bottle[] {
  try {
    const raw = getItem(LS_BOTTLES);
    const parsed = raw ? (JSON.parse(raw) as Bottle[]) : [];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

export function saveBottles(b: Bottle[]): void {
  setItem(LS_BOTTLES, JSON.stringify(b));
}

/* ── Expenses ── */
export const LS_EXPENSES = "gasrefill_expenses_v1";

export function loadExpenses(): Expense[] {
  try {
    const raw = getItem(LS_EXPENSES);
    const parsed = raw ? (JSON.parse(raw) as Expense[]) : [];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

export function saveExpenses(e: Expense[]): void {
  setItem(LS_EXPENSES, JSON.stringify(e));
}

/* ── Business Settings ── */
export const LS_SETTINGS = "gasrefill_settings_v1";

export function loadSettings(): BusinessSettings | null {
  try {
    const raw = getItem(LS_SETTINGS);
    return raw ? (JSON.parse(raw) as BusinessSettings) : null;
  } catch {
    return null;
  }
}

export function saveSettings(s: BusinessSettings): void {
  setItem(LS_SETTINGS, JSON.stringify(s));
}

/* ── Cost Price ── */
export const LS_COST_PRICE = "gasrefill_cost_price_v1";

export function loadCostPrice(fallback: string): string {
  const v = getItem(LS_COST_PRICE);
  return v !== null && v !== "" ? v : fallback;
}

export function saveCostPrice(v: string): void {
  setItem(LS_COST_PRICE, v);
}

/* ── Backup & Restore ── */
export function exportBackupData(): BackupData {
  return {
    version: 1,
    exportedAt: Date.now(),
    businessName: "",
    transactions: loadHistory(),
    bottles: loadBottles(),
    expenses: loadExpenses(),
    settings: loadSettings() ?? {
      businessName: "Gas Refill Terminal",
      receiptFooter: "Thank you for your patronage!",
      currencySymbol: "\u20A6",
      currencyCode: "NGN",
    },
  };
}

export function importBackupData(data: BackupData): {
  transactions: number;
  bottles: number;
  expenses: number;
} {
  if (!data || data.version !== 1) {
    throw new Error("Invalid backup file format");
  }
  if (Array.isArray(data.transactions)) saveHistory(data.transactions);
  if (Array.isArray(data.bottles)) saveBottles(data.bottles);
  if (Array.isArray(data.expenses)) saveExpenses(data.expenses);
  if (data.settings) saveSettings(data.settings);

  return {
    transactions: data.transactions?.length ?? 0,
    bottles: data.bottles?.length ?? 0,
    expenses: data.expenses?.length ?? 0,
  };
}

export function clearAllData(): void {
  removeItem(LS_GSP);
  removeItem(LS_HIST);
  removeItem(LS_BOTTLES);
  removeItem(LS_EXPENSES);
  removeItem(LS_SETTINGS);
  removeItem(LS_COST_PRICE);
}

File: android-app/app/src/main/java/com/gasrefill/terminal/utils/Storage.kt
package com.gasrefill.terminal.utils

import android.content.Context
import android.content.SharedPreferences
import com.gasrefill.terminal.models.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

object Storage {
    private const val PREFS_NAME = "gasrefill_prefs"
    private const val KEY_GSP = "gsp"
    private const val KEY_HISTORY = "history"
    private const val KEY_BOTTLES = "bottles"
    private const val KEY_EXPENSES = "expenses"
    private const val KEY_SETTINGS = "settings"
    private const val KEY_COST_PRICE = "cost_price"
    private const val KEY_THEME = "theme"

    private val json = Json {
        ignoreUnknownKeys = true
        prettyPrint = true
    }

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    // GSP
    fun loadGsp(context: Context, fallback: String = "1700"): String {
        return prefs(context).getString(KEY_GSP, fallback) ?: fallback
    }

    fun saveGsp(context: Context, value: String) {
        prefs(context).edit().putString(KEY_GSP, value).apply()
    }

    // History
    fun loadHistory(context: Context): List<Transaction> {
        val raw = prefs(context).getString(KEY_HISTORY, null) ?: return emptyList()
        return try {
            json.decodeFromString<List<Transaction>>(raw)
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun saveHistory(context: Context, transactions: List<Transaction>) {
        prefs(context).edit().putString(KEY_HISTORY, json.encodeToString(transactions)).apply()
    }

    // Bottles
    fun loadBottles(context: Context): List<Bottle> {
        val raw = prefs(context).getString(KEY_BOTTLES, null) ?: return emptyList()
        return try {
            json.decodeFromString<List<Bottle>>(raw)
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun saveBottles(context: Context, bottles: List<Bottle>) {
        prefs(context).edit().putString(KEY_BOTTLES, json.encodeToString(bottles)).apply()
    }

    // Expenses
    fun loadExpenses(context: Context): List<Expense> {
        val raw = prefs(context).getString(KEY_EXPENSES, null) ?: return emptyList()
        return try {
            json.decodeFromString<List<Expense>>(raw)
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun saveExpenses(context: Context, expenses: List<Expense>) {
        prefs(context).edit().putString(KEY_EXPENSES, json.encodeToString(expenses)).apply()
    }

    // Settings
    fun loadSettings(context: Context): BusinessSettings? {
        val raw = prefs(context).getString(KEY_SETTINGS, null) ?: return null
        return try {
            json.decodeFromString<BusinessSettings>(raw)
        } catch (e: Exception) {
            null
        }
    }

    fun saveSettings(context: Context, settings: BusinessSettings) {
        prefs(context).edit().putString(KEY_SETTINGS, json.encodeToString(settings)).apply()
    }

    // Cost Price
    fun loadCostPrice(context: Context, fallback: String = "0"): String {
        return prefs(context).getString(KEY_COST_PRICE, fallback) ?: fallback
    }

    fun saveCostPrice(context: Context, value: String) {
        prefs(context).edit().putString(KEY_COST_PRICE, value).apply()
    }

    // Theme
    fun loadTheme(context: Context): String? = prefs(context).getString(KEY_THEME, null)
    fun saveTheme(context: Context, theme: String) {
        prefs(context).edit().putString(KEY_THEME, theme).apply()
    }

    // Clear all
    fun clearAll(context: Context) {
        prefs(context).edit().clear().apply()
    }
}
